### ISSUE TITLE
Runtime exception when using .NET 5.0 projection generated from C++/WinRT in WinUI 3 Preview 3 desktop app

### ISSUE THREAD

**Issue by Rochishaa** [microsoft/CsWinRT#653]
In my current UWP project, I was using C++/WinRT components. The registration there was done automatically if I added the reference to the .winmd. I am looking at migrating the same app to WinUI 3.

For starters, I am trying to consume SimpleMathComponent C++/WinRT component in a sample WinUI 3.0 Preview 3 Desktop app. Following the steps as described [here](https://docs.microsoft.com/en-us/windows/uwp/csharp-winrt/net-projection-from-cppwinrt-component), I created .NET 5.0 projection and added the nuget package as reference to my sample winui for desktop app. Build is successful, but still getting runtime exception. The reason being missing registration. 

I now manually added registration to the app.manifest in WinUI sample app. 

This is what my WinUI app.manifest looks like : 

`<?xml version="1.0" encoding="utf-8"?>
<assembly manifestVersion="1.0" xmlns="urn:schemas-microsoft-com:asm.v1">
  <assemblyIdentity version="1.0.0.0" name="App1.app"/>
  <file name="SimpleMathProjection.dll">
    <activatableClass
    name="SimpleMathComponent.SimpleMath"
    threadingModel="both"
    xmlns="urn:schemas-microsoft-com:winrt.v1" />
  </file>
</assembly>`

This is the exception I am getting this time around:

System.Reflection.TargetInvocationException
  HResult=0x80131604
  Message=Exception has been thrown by the target of an invocation.
  Source=System.Private.CoreLib
  StackTrace:
   at System.RuntimeTypeHandle.CreateInstance(RuntimeType type, Boolean publicOnly, Boolean wrapExceptions, Boolean& canBeCached, RuntimeMethodHandleInternal& ctor, Boolean& hasNoDefaultCtor)
   at System.RuntimeType.CreateInstanceDefaultCtorSlow(Boolean publicOnly, Boolean wrapExceptions, Boolean fillCache)
   at System.RuntimeType.CreateInstanceDefaultCtor(Boolean publicOnly, Boolean skipCheckThis, Boolean fillCache, Boolean wrapExceptions)
   at System.Activator.CreateInstance[T]()
   at WinRT.WeakLazy`1.get_Value()
   at WinRT.ActivationFactory`1.ActivateInstance[I]()
   at SimpleMathComponent.SimpleMath..ctor()
   at App1.MainWindow.myButton_Click(Object sender, RoutedEventArgs e) in C:\Dev\CsWinRt\src\Samples\Net5ProjectionSample\App1\App1\MainWindow.xaml.cs:line 35
   at ABI.Microsoft.UI.Xaml.RoutedEventHandler.<>c__DisplayClass10_0.<Do_Abi_Invoke>b__0(RoutedEventHandler invoke)
   at WinRT.ComWrappersSupport.MarshalDelegateInvoke[T](IntPtr thisPtr, Action`1 invoke)
   at ABI.Microsoft.UI.Xaml.RoutedEventHandler.Do_Abi_Invoke(IntPtr thisPtr, IntPtr sender, IntPtr e)

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
COMException: Error in the DLL (0x800401F9 (CO_E_ERRORINDLL))

_Originally posted by @Rochishaa in https://github.com/microsoft/CsWinRT/issues/647#issuecomment-754812064_

**Comment by AdamBraden** [comment]
@Rochishaa - are you trying to run the Packaging project or the Exe project? If you are running the exe project - I suspect you are missing the VCLibs framework package that the c++ winrt component depends on. You may need to add a reference to the VCRT Forwarders nuget package, and then ensure you have the appropriate flavor of VCRedist installed (debug or release). If you share a repo it will be easier to confirm.

**Comment by manodasanW** [comment]
@Rochishaa In addition to what Adam mentioned, I noticed in your app.manifest, you have SimpleMathProjection.dll as the DLL.  I am assuming this is the CsWinRT projection DLL for your SimpleMath component.  If so, you actually in the manifest want to specify your component DLL itself (i.e. SimpleMathComponent.dll or however it is named) and not the projection DLL.

**Comment by Rochishaa** [comment]
@AdamBraden - I am running the packaging project. The projection nuget package is added to the exe project. It has got the VCLibs framework. I have uploaded my code [here](https://github.com/Rochishaa/CCinWinUI3)

**Comment by Rochishaa** [comment]
@manodasanW - If I specify the SimpleMathComponent.dll in the manifest, I get "FileNotFoundException: The specified module could not be found. (0x8007007E)"

**Comment by angelazhangmsft** [comment]
@Rochishaa I noticed in your .nuspec file is missing the file SimplemathComponent.dll like the [sample](https://github.com/microsoft/CsWinRT/blob/29d8ec05654b6b639f578acabbe4405a69108221/src/Samples/Net5ProjectionSample/SimpleMathProjection/nuget/SimpleMathProjection.nuspec#L23). The nuget package should deploy both SimpleMathComponent.dll (implementation) and SimpleMathProjection.dll.  If the nuget package is intended to be used by both native and managed apps, then the nuget would further need to filter the file entries based on TPM

**Comment by Rochishaa** [comment]
@angelazhangmsft When I add 
<file _src="..\..\_build\x64\Debug\SimpleMathComponent\bin\SimpleMathComponent\SimpleMathComponent.dll" target="runtimes\win10-x64\native\SimpleMathComponent.dll" />_
to my nuspec. I get a build error saying 
Error: The given key was not present in the dictionary.	
Project: SimpleMathProjection	
File: C:\Program Files\dotnet\sdk\5.0.200-preview.20601.7\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets	
Line: 221

**Comment by angelazhangmsft** [comment]
@Rochishaa - that seems to be a bug with that .NET SDK preview version of 5.0.200-preview.20601.7. I was able to successfully build with 5.0.200-preview.21056.17 (or you could revert to the latest release, 5.0.101). After that I was able to use SimpleMathComponent in a WinUI3 Desktop app without the need for registration and just with the VCRT forwarders package.

I recommend staying with VS 16.9 Preview, as that version is required for WinUI3 Preview 3. Feel free to reopen if this does not solve your issue.

**Comment by Rochishaa** [comment]
@angelazhangmsft  - I managed to install 5.0.200-preview.21056.17. The nuspec is building without any issues now,. But I am still getting winrt class registration issue for SimpleMathComponent. 

FYI, I found no way to uninstall 5.0.200-preview.20601.7. It gets installed as part of VS 16.9.0 Preview 2.0 and I couldn't find the option to uninstall it.

**Comment by Rochishaa** [comment]
@angelazhangmsft - I set up another machine and somehow I was able to use SimpleMathComponent in a WinUI3 Desktop app without the need for registration and just with the VCRT forwarders package. 

Here are the setup configs of the 2 machines:

- Machine 1 : Throwing runtime exception when using .NET 5.0 projection generated from C++/WinRT in WinUI 3 Preview 3 desktop app

       1.1 OS : 1909
       1.2 VS : 16.9.0 Preview 2.0
       1.3 SDK : 5.0.200-preview.21057.7

- Machine 2 : Able to use SimpleMathComponent in a WinUI3 Desktop app without the need for registration.
       1.1 OS : 1809
       1.2 VS : 16.9.0 Preview 2.0
       1.3 SDK : 5.0.200-preview.21060.4

**Comment by Rochishaa** [comment]
@angelazhangmsft - Tried a couple of things to troubleshoot. Finally, here is what worked for me:

1. Installed 20H2. 
2. Installed VS 16.9.0 Preview 2.0.  .NET SDK 5.0.200-preview.20601.7 gets installed with it.
3. Build error for .nuspec in SimpleMathProjection
```
<file _src="..\..\_build\x64\Debug\SimpleMathComponent\bin\SimpleMathComponent\SimpleMathComponent.dll" target="runtimes\win10-x64\native\SimpleMathComponent.dll" />

Error: The given key was not present in the dictionary.
Project: SimpleMathProjection
File: C:\Program Files\dotnet\sdk\5.0.200-preview.20601.7\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets
Line: 221
```
4. Install .NET SDK 5.0.200-preview.21062.15. The build error is gone.
5. Able to use SimpleMathComponent in a WinUI3 Desktop app without the need for registration.

**Comment by angelazhangmsft** [comment]
Hi @Rochishaa - it seems like you're unblocked. I took a look at your repo again and I didn't have any problems building (I'm on a 2004 OS version) . Was the registration problem with different OS versions only occurring with the WinUI3 app, or also the console app? 

OS version shouldn't affect registration, so it must be something else in the environment. You could use procmon to observe dll name/path issues, sxstrace to observe reg-free activation errors if you wanted to look further.

**Comment by Rochishaa** [comment]
@angelazhangmsft - You are right. It looks like it was something else in the environment. I was getting the registration problem with the console app also.

**Comment by Scottj1s** [comment]
@Rochishaa - it looks like most of the registration issues originally reported in this issue have been addressed.  Please open another issue if problems remain.
